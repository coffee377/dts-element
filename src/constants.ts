import * as ts from 'typescript';
import {create_native_type, INativeType} from './types/native-type';

export enum ElementKind {
  GenericDeclaration,
  GenericType,
  NativeType,
  ArrayType,
  Parameter,
}

export const any_type = create_native_type({type: ts.createKeywordTypeNode(ts.SyntaxKind.AnyKeyword)});
export const never_type = create_native_type({type: ts.createKeywordTypeNode(ts.SyntaxKind.NeverKeyword)});

export const null_type = create_native_type({type: ts.createKeywordTypeNode(ts.SyntaxKind.NullKeyword)});
export const undefined_type = create_native_type({type: ts.createKeywordTypeNode(ts.SyntaxKind.UndefinedKeyword)});
export const void_type = create_native_type({type: ts.createKeywordTypeNode(ts.SyntaxKind.VoidKeyword)});

export const boolean_type = create_native_type({type: ts.createKeywordTypeNode(ts.SyntaxKind.BooleanKeyword)});
export const number_type = create_native_type({type: ts.createKeywordTypeNode(ts.SyntaxKind.NumberKeyword)});
export const string_type = create_native_type({type: ts.createKeywordTypeNode(ts.SyntaxKind.StringKeyword)});
export const symbol_type = create_native_type({type: ts.createKeywordTypeNode(ts.SyntaxKind.SymbolKeyword)});
export const object_type = create_native_type({type: ts.createKeywordTypeNode(ts.SyntaxKind.ObjectKeyword)});

export const this_type = create_native_type({type: ts.createKeywordTypeNode(ts.SyntaxKind.ThisKeyword)});
