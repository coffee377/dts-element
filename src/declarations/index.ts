export * from './class-declaration';
export * from './function-declaration';
export * from './generic-declaration';
export * from './interface-declaration';
export * from './parameter-declaration';
export * from './type-declaration';
export * from './variable-declaration';
