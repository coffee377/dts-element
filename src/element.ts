export abstract class Element<Parameters extends {}> {

  public parameters: Parameters;

  constructor(parameters: Parameters) {
    this.parameters = parameters;
  }

  public get<Key extends keyof Parameters>(key: Key): Parameters[Key] {
    return this.parameters[key];
  }

  public set(values: Partial<Parameters>): this {
    Object.keys(values).forEach((key: keyof Parameters) => {
      const value = values[key];
      if (value !== undefined) {
        this.parameters[key] = value;
      }
    });
    return this;
  }

  public clone(is_deep_clone: boolean = false): this {
    const element = Object.create(this.constructor.prototype);
    element.parameters = {};
    element.copy(this, is_deep_clone);
    return element;
  }

  public copy(element: this, is_deep_copy: boolean = false): this {
    this.parameters = this._clone_parameters(element.parameters, is_deep_copy);
    return this;
  }

  private _clone_parameters(parameters: Parameters, is_deep_clone: boolean): Parameters {
    return Object.keys(parameters).reduce((cloned_parameters: Partial<Parameters>, key: keyof Parameters) => {
      const value = parameters[key];
      const cloned_value = this._clone_value(value, is_deep_clone);
      cloned_parameters[key] = cloned_value;
      return cloned_parameters;
    }, {} as any);
  }

  private _clone_value<T>(value: T, is_deep_clone: boolean): T {
    return (typeof value !== 'object')
      ? value
      : (value instanceof Element)
        ? value.clone(is_deep_clone)
        : (value instanceof Array)
          ? value.map(<U>(x: U): U => this._clone_value(x, is_deep_clone)) as any
          : value; // TODO
  }

}
