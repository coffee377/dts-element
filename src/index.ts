export * from './collections';
export * from './constants';
export * from './declarations/index';
export * from './element';
export * from './emit';
export * from './transform';
export * from './types/index';
