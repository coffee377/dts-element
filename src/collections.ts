import * as declarations from './declarations/index';
import * as types from './types/index';

export type IDeclaration =
  | declarations.IFunctionDeclaration
  | declarations.IGenericDeclaration
  | declarations.IParameterDeclaration
  | declarations.IVariableDeclaration
;

export type IType =
  | types.IArrayType
  | types.IConstructorType
  | types.IFunctionType
  | types.IGenericType
  | types.IIntersectionType
  | types.IKeyofType
  | types.ILiteralType
  | types.INativeType
  | types.ITypeofType
  | types.ITupleType
  | types.IUnionType
;
