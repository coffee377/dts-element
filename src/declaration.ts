import {Element} from './element';
import {line_map} from './helpers/line-map';

export interface IDeclarationRequiredParameters {
  name: string;
}

export interface IDeclarationOptionalParameters {
  jsdoc: string;
}

export type AnyDeclaration = Declaration<any, any>;

export abstract class Declaration<RequiredParameters extends {}, OptionalParameters extends {}> extends Element
    <IDeclarationRequiredParameters & RequiredParameters, IDeclarationOptionalParameters & OptionalParameters> {

  public get default_declaration_parameters(): IDeclarationOptionalParameters {
    return {
      jsdoc: '',
    };
  }

  public emit_jsdoc(): string {
    const {jsdoc} = this.parameters;
    const content = line_map(jsdoc, (line: string) => ` * ${line}`);
    return (content.length === 0)
      ? content
      : `/**\n${content}\n */\n`;
  }

}
