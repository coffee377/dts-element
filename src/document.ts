import {Element} from './element';
import {RootElement} from './sets';
import {trim_every_line} from './helpers/trim-every-line';
import {AnyElement} from './element';

export interface IDocumentRequiredParameters {
  elements: RootElement[];
}

// tslint:disable-next-line no-empty-interface
export interface IDocumentOptionalParameters {}

export class Document extends Element<IDocumentRequiredParameters, IDocumentOptionalParameters> {

  public get default_parameters(): IDocumentOptionalParameters {
    return {};
  }

  public _emit(_container: AnyElement): string {
    const {elements} = this.parameters;
    const content = elements.map((element: RootElement) => element.emit()).join('\n');
    return trim_every_line(content);
  }

}
