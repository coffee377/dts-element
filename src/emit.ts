import * as ts from 'typescript';
import {ElementKind} from './constants';
import {IElement} from './element';
import {emit_triple_slash_reference, ITripleSlashReference} from './others/triple-slash-reference';
import {transform} from './transform';

export const emit = (element: IElement<any>, options?: ts.PrinterOptions) => {
  switch (element.kind) {
    case ElementKind.TripleSlashReference:
      return emit_triple_slash_reference(element as ITripleSlashReference);
    default:
      const ts_node = transform(element);
      const source_file = ts.createSourceFile('', '', ts.ScriptTarget.Latest, false, ts.ScriptKind.TS);
      const printer = ts.createPrinter(options);
      return printer.printNode(ts.EmitHint.Unspecified, ts_node, source_file);
  }
};
