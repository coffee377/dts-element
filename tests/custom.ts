jest.disableAutomock();

import * as dts from '../src/index';

const parameter_assertion_a = new dts.Parameter({
  name: 'a',
  type: dts.any_type,
});

const function_assertion = new dts.FunctionDeclaration({
  name: 'assertion',
  parameters: [parameter_assertion_a],
  return: new dts.TypeAssertion({
    parameter: parameter_assertion_a,
    type: dts.string_type,
  }),
});

const reference_somepath = new dts.Reference({
  kind: 'path',
  content: 'path/to/some.d.ts',
});

const object_type_generics = new dts.ObjectType({
  members: [
    new dts.ObjectMember({
      owned: ((
        generic_type_T: dts.GenericType,
        generic_type_U: dts.GenericType,
      ): dts.FunctionDeclaration => new dts.FunctionDeclaration({
        name: 'a',
        generics: [
          generic_type_T,
          generic_type_U,
        ],
        return: new dts.UnionType({
          types: [generic_type_T, generic_type_U],
        }),
      }))(
        new dts.GenericType({name: 'T'}),
        new dts.GenericType({name: 'U'}),
      ),
    }),
    new dts.ObjectMember({
      owned: new dts.IndexSignature({
        name: 'index',
        type: dts.any_type,
      }),
    }),
  ],
});

const type_declaration_t1 = new dts.TypeDeclaration({
  name: 't1',
  target: object_type_generics,
});

const enum_Flags = new dts.EnumDeclaration({
  name: 'Flags',
  members: [
    new dts.EnumMember({
      owned: new dts.VariableDeclaration({name: 'A'}),
    }),
    new dts.EnumMember({
      owned: new dts.VariableDeclaration({name: 'B'}),
      value: -2,
    }),
    new dts.EnumMember({
      owned: new dts.VariableDeclaration({name: 'C'}),
    }),
  ],
});

const export_all = new dts.ExportAll({from: 'path/to/other'});
const export_default = new dts.ExportDefault({value: type_declaration_t1});

const import_all = new dts.ImportAll({from: 'path/to/another'});

const variable_declaration_imported = new dts.VariableDeclaration({name: 'imported'});
const import_default = new dts.ImportDefault({
  from: 'path/to/something',
  value: variable_declaration_imported,
});

const variable_declaration_imported_named_1 = new dts.VariableDeclaration({name: 'imported_named_1'});
const variable_declaration_imported_named_2 = new dts.VariableDeclaration({name: 'imported_named_2'});
const import_named = new dts.ImportNamed({
  from: 'path/to/abc',
  members: [
    new dts.ImportMember({
      owned: variable_declaration_imported_named_1,
      name: 'some_1',
    }),
    new dts.ImportMember({
      owned: variable_declaration_imported_named_2,
      name: 'some_2',
    }),
  ],
});

const export_named = new dts.ExportNamed({
  members: [
    new dts.ExportMember({
      owned: variable_declaration_imported_named_1,
      name: 'out_1',
    }),
    new dts.ExportMember({
      owned: variable_declaration_imported_named_2,
      name: 'out_2',
    }),
  ],
});

const intersection_type = new dts.IntersectionType({
  types: [
    dts.string_type,
    new dts.KeyofType({owned: object_type_generics}),
    new dts.TypeofType({owned: variable_declaration_imported}),
    new dts.LiteralType({value: 'literal'}),
    new dts.TupleType({owneds: [
      dts.string_type,
      dts.null_type,
    ]}),
  ],
});

const type_declaration_t2 = new dts.TypeDeclaration({
  name: 't2',
  export: true,
  target: intersection_type,
});

const class_Element = new dts.ClassDeclaration({
  name: 'Element',
  export: true,
  members: [
    new dts.ClassMember({
      owned: new dts.VariableDeclaration({name: 'parameters'}),
    }),
    new dts.ClassMember({
      owned: new dts.FunctionDeclaration({
        name: 'test',
        return: new dts.SubType({
          target: dts.this_type,
          path: [new dts.LiteralType({value: 'parameters'})],
        }),
      }),
    }),
  ],
});

const document = new dts.Document({children: [
  reference_somepath,
  import_all,
  import_default,
  import_named,
  function_assertion,
  class_Element,
  type_declaration_t1,
  enum_Flags,
  export_all,
  export_named,
  export_default,
  type_declaration_t2,
]});

it('should return correctly', () => {
  expect(document.emit()).toMatchSnapshot();
});
