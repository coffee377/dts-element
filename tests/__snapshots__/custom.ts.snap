// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return correctly 1`] = `
"/// <reference path=\\"path/to/some.d.ts\\" />
import \\"path/to/another\\";
import imported from \\"path/to/something\\";
import {some_1 as imported_named_1, some_2 as imported_named_2} from \\"path/to/abc\\";
declare function assertion(a: any): a is string;
export class Element {
  parameters: any;
  test(): this[\\"parameters\\"];
}
type t1 = {
  a<T, U>(): T | U;
  [index: string]: any;
};
declare enum Flags {
  A = 0,
  B = -2,
  C = -1
}
export * from \\"path/to/other\\";
export {imported_named_1 as out_1, imported_named_2 as out_2};
export default t1;
export type t2 = string & (keyof {
  a<T, U>(): T | U;
  [index: string]: any;
}) & (typeof imported) & \\"literal\\" & [string, null];"
`;
